src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs

```diff
@@ -196,8 +199,27 @@ private static MethodSymbol GetEntryPoint(CSharpCompilation compilation, PEModul
             diagnostics.AddRange(entryPointAndDiagnostics.Diagnostics);
 
             var entryPoint = entryPointAndDiagnostics.MethodSymbol;
-            var synthesizedEntryPoint = entryPoint as SynthesizedEntryPointSymbol;
-            if (((object)synthesizedEntryPoint != null) &&
+
+            if (entryPoint == null)
+            {
+                Debug.Assert(entryPointAndDiagnostics.Diagnostics.HasAnyErrors() || !compilation.Options.Errors.IsDefaultOrEmpty);
+                return null;
+            }
+
+            SynthesizedEntryPointSymbol synthesizedEntryPoint = null;
+            bool addedDefinition = false;
+            if (entryPoint is SynthesizedEntryPointSymbol s)
+            {
+                synthesizedEntryPoint = s;
+            }
+            else if (entryPoint.HasAsyncMainReturnType(compilation) && compilation.LanguageVersion >= LanguageVersion.CSharp7_1)
+            {
+                synthesizedEntryPoint = new AsyncForwardEntryPoint(compilation, diagnostics, entryPoint.ContainingType, entryPoint);
```

Comment:
Right now the only place that the diagnosticsBag is used is during the `GetRequiredMethod` lookup for `GetAwaiter` and `GetResult`.